braindecode/models/eegconformer.py

```diff
@@ -0,0 +1,313 @@
+        att = F.softmax(energy / scaling, dim=-1)
+        att = self.att_drop(att)
+        out = torch.einsum("bhal, bhlv -> bhav ", att, values)
+        out = rearrange(out, "b h n d -> b n (h d)")
+        out = self.projection(out)
+        return out
+
+
+class _ResidualAdd(nn.Module):
+    def __init__(self, fn):
+        super().__init__()
+        self.fn = fn
+
+    def forward(self, x, **kwargs):
+        res = x
+        x = self.fn(x, **kwargs)
+        x += res
+        return x
+
+
+class _FeedForwardBlock(nn.Sequential):
+    def __init__(self, emb_size, expansion, drop_p):
+        super().__init__(
+            nn.Linear(emb_size, expansion * emb_size),
+            nn.GELU(),
+            nn.Dropout(drop_p),
+            nn.Linear(expansion * emb_size, emb_size),
+        )
+
+
+class _TransformerEncoderBlock(nn.Sequential):
+    def __init__(self, emb_size, att_heads, att_drop, forward_expansion=4):
+        super().__init__(
+            _ResidualAdd(
+                nn.Sequential(
+                    nn.LayerNorm(emb_size),
+                    _MultiHeadAttention(emb_size, att_heads, att_drop),
+                    nn.Dropout(att_drop),
+                )
+            ),
+            _ResidualAdd(
+                nn.Sequential(
+                    nn.LayerNorm(emb_size),
+                    _FeedForwardBlock(
+                        emb_size, expansion=forward_expansion,
+                        drop_p=att_drop
+                    ),
+                    nn.Dropout(att_drop),
+                )
+            ),
+        )
+
+
+class _TransformerEncoder(nn.Sequential):
+    """Transformer encoder module for the transformer encoder.
+
+    Similar to the layers used in ViT.
+    """
+    def __init__(self, att_depth, emb_size, att_heads, att_drop):
+        super().__init__(
+            *[
+                _TransformerEncoderBlock(emb_size, att_heads, att_drop)
+                for _ in range(att_depth)
+            ]
+        )
+
+
+class ClassificationHead(nn.Module):
+    def __init__(self, emb_size, final_fc_length, n_classes,
+                 drop_prob_1=0.5, drop_prob_2=0.3, out_channels=256,
+                 hidden_channels=32, return_features=False):
+        """"Classification head for the transformer encoder.
+
+        Parameters
+        ----------
+        emb_size : int
+            Embedding size of the transformer encoder.
+        final_fc_length : int
+            Length of the final fully connected layer.
+        n_classes : int
+            Number of classes for classification.
+        drop_prob_1 : float
+            Dropout probability for the first dropout layer.
+        drop_prob_2 : float
+            Dropout probability for the second dropout layer.
+        out_channels : int
+            Number of output channels for the first linear layer.
+        hidden_channels : int
+            Number of output channels for the second linear layer.
+        """
```

Comment:
Missing docstring for `return_features` - do we use that anywhere?
```suggestion
        return_features : bool
            Whether to return input features.
        """
```